lte/gateway/c/session_manager/ShardTracker.cpp

```diff
@@ -13,45 +13,58 @@
 
 #include <string>
 #include <vector>
+#include <set>
 
 #include "ShardTracker.h"
 
 namespace magma {
 
-int ShardTracker::add_ue() {
-  /*
-  * If there are no shards, add an entry for the shard
-  * representing a single UE at the new shard
-  */
-  if (ue_count_per_shard_.size() == 0) {
-    ue_count_per_shard_.push_back(1);
-    return 0;
+int ShardTracker::add_ue(std::string imsi) {
+  /**
+   * If there are no shards, add a single entry set
+   * to shards list containing the IMSI
+   * */
+  if (shards_.empty()) {
+    std::set<std::string> new_shard = {imsi};
+    shards_.push_back(new_shard);
   }
-  /*
-  * Iterate through all existing shards, if all are full
-  * create a new shard with quantity 1, otherwise increment
-  * the quantity of UEs in the latest shard
-  */
-  for (size_t shard_id = 0; shard_id < ue_count_per_shard_.size(); shard_id++) {
-    if (ue_count_per_shard_[shard_id] < max_shard_size_) {
-      ue_count_per_shard_[shard_id]++;
+
+  for (size_t shard_id = 0; shard_id < shards_.size(); shard_id++) {
+    /**
+     * If the UE is already in the shard, return the shard id. This check
+     * is meant to avoid multiple sessions for a UE being assigned duplicate
+     * entries.
+     * */
+    if (shards_[shard_id].find(imsi) != shards_[shard_id].end()) {
+      return shard_id;
```

Comment:
Aren't we storing shard id on session_map. If the UE already exists on session_map, we can get the shard id from there

In case we do not store anymore. Does it means every time we have to add an shard we have to go through all the shards in the systen?

Reply:
That's a good point I can simplify this by just accessing it directly. For the second question, the answer is yes. I mentioned in the PR description that I still need to implement state recovery, and to do that I'd need to go through all the shards.

Reply:
I ran into some errors that I could not figure out for some reason when trying to access the shard id directly. Since this approaches work and isn't too much more inefficient(O(number of shards), and we'll probably use very few shards(as discussed in our design doc), I am going to keep this check in for now, but it'll be addressed in a future PR(probably the state recovery PR).